proof/refine/AARCH64/ArchCSpace_I.thy

```diff
@@ -737,416 +183,91 @@ lemma sameObjectAs_def2:
-  "capUntypedSize (EndpointCap r badge a b c d) = 1 << objBits (undefined::endpoint)"
-  "capUntypedSize (Zombie r zs n) = 1 << (zBits zs)"
-  "capUntypedSize NullCap = 0"
-  "capUntypedSize DomainCap = 1"
-  "capUntypedSize (ArchObjectCap x) = Arch.capUntypedSize x"
-  "capUntypedSize (UntypedCap d r n f) = 1 << n"
-  "capUntypedSize (CNodeCap r n g n2) = 1 << (objBits (undefined::cte) + n)"
-  "capUntypedSize (ReplyCap r m a) = 1 << objBits (undefined :: tcb)"
-  "capUntypedSize IRQControlCap = 1"
-  "capUntypedSize (IRQHandlerCap irq) = 1"
-  by (auto simp add: capUntypedSize_def isCap_simps objBits_simps'
-              split: zombie_type.splits)
-
-lemma sameRegionAs_classes:
+lemma sameRegionAs_classes[CSpace_I_2_assms]:
   "sameRegionAs cap cap' \<Longrightarrow> capClass cap = capClass cap'"
-  apply (erule sameRegionAsE)
-     apply (rule master_eqI, rule capClass_Master)
-     apply simp
-    apply (simp add: capRange_def split: if_split_asm)
-   apply (clarsimp simp: isCap_simps)+
-  done
-
-lemma capAligned_capUntypedPtr:
-  "\<lbrakk> capAligned cap; capClass cap = PhysicalClass \<rbrakk> \<Longrightarrow>
-   capUntypedPtr cap \<in> capRange cap"
-  by (simp add: capRange_def capAligned_def is_aligned_no_overflow)
+  by (erule sameRegionAsE, rule master_eqI)
+     (clarsimp simp: capRange_def isCap_simps intro!: capClass_Master split: if_split_asm)+
 
-lemma sameRegionAs_capRange_Int:
+lemma sameRegionAs_capRange_Int[CSpace_I_2_assms]:
   "\<lbrakk> sameRegionAs cap cap'; capClass cap = PhysicalClass \<or> capClass cap' = PhysicalClass;
      capAligned cap; capAligned cap' \<rbrakk>
-     \<Longrightarrow> capRange cap' \<inter> capRange cap \<noteq> {}"
+   \<Longrightarrow> capRange cap' \<inter> capRange cap \<noteq> {}"
   apply (frule sameRegionAs_classes, simp)
   apply (drule(1) capAligned_capUntypedPtr)+
   apply (erule sameRegionAsE)
       apply (subgoal_tac "capRange (capMasterCap cap') \<inter> capRange (capMasterCap cap) \<noteq> {}")
        apply (simp(no_asm_use) add: capRange_Master)
-      apply (clarsimp simp: capRange_Master)
-     apply blast
-    apply blast
-   apply (clarsimp simp: isCap_simps)
-  apply (clarsimp simp: isCap_simps)
+      apply (fastforce simp: capRange_Master isCap_simps)+
   done
 
 lemma sameRegionAs_trans:
   "\<lbrakk> sameRegionAs a b; sameRegionAs b c \<rbrakk> \<Longrightarrow> sameRegionAs a c"
-  apply (simp add: sameRegionAs_def2, elim conjE disjE, simp_all)
-                by (auto simp: isCap_simps capRange_def) (* long *)
+  by (simp add: sameRegionAs_def2, elim conjE disjE)
+     (auto simp: isCap_simps capRange_def) (* long *)
 
+(* FIXME arch-split: TODO interface *)
 lemma capMasterCap_maskCapRights[simp]:
-  "capMasterCap (maskCapRights msk cap)
-    = capMasterCap cap"
-  apply (cases cap;
-         simp add: maskCapRights_def Let_def isCap_simps capMasterCap_def)
+  "capMasterCap (maskCapRights msk cap) = capMasterCap cap"
```

Comment:
Is this a FIXME that's resolvable now or do you want to come back to it later when the use-cases are better defined?

Reply:
I probably should've scrapped all these `TODO interface` FIXMEs, but it does raise why I got a bit stuck on them. This is a lemma that could in theory be used outside of Arch, but it doesn't seem to right now. So the obvious thing is to not make it an interface lemma right now... but on the other hand, for new lemmas/theories/etc in the future, it then won't be in scope for people to use. I'm a bit torn on the issue. What do you think?

Reply:
Hmm, that's a tough one. I think I lean towards adding lemmas like this that you spot to the interface. It's a bit more boilerplate now but could avoid a lot of pain in the future if they do end up being needed.

Reply:
Haha, after making more progress in reviewing this, looks like you already did add it to the interface and remove this FIXME in a later commit.

Reply:
Damn, did I squash those into the wrong commit?